### PR TITLE
Jest framework

### PR DESCRIPTION
**Pull request by bdemann**

### CODE REVIEW

File: .github/workflows/test.yml [hunk 1]
```diff
@@ -321,7 +321,7 @@ jobs:
             - if: ${{ needs.release-candidate-deploy.outputs.should_run_tests && !contains(github.head_ref, 'release--') && !(github.ref == 'refs/heads/main' && contains(github.event.head_commit.message, 'Merge pull request') && contains(github.event.head_commit.message, 'demergent-labs/release--')) }}
               shell: bash -l {0}
               working-directory: ${{ matrix.example_directories }}
-              run: AZLE_PROPTEST_NUM_RUNS=5 AZLE_PROPTEST_VERBOSE=true npm test
+              run: AZLE_PROPTEST_NUM_RUNS=5 AZLE_PROPTEST_VERBOSE=true AZLE_INTEGRATION_TEST_LINK_AZLE=true npm test
```
Comment (bdemann) [inline .github/workflows/test.yml:324]: We should discuss this during the review

File: examples/motoko_examples/superheroes/src/declarations/index.d.ts
Comment (bdemann) [inline examples/motoko_examples/superheroes/src/declarations/index.d.ts:1]: Why do we have the declaration files for superheroes and phone_book in the repo? Is that intentional?

File: examples/basic_bitcoin/test/bitcoin_daemon.ts [hunk 2]
```diff
@@ -1,15 +1,9 @@
...
 import { existsSync, rmSync } from 'fs-extra';
 
-export async function whileRunningBitcoinDaemon(
-    callback: () => Promise<boolean> | void
-) {
-    const bitcoinDaemon = await startBitcoinDaemon();
-    await callback();
-    bitcoinDaemon.kill();
-}
+export { ChildProcessWithoutNullStreams as BitcoinDaemon };
```
Comment (lastmjs) [inline examples/basic_bitcoin/test/bitcoin_daemon.ts:4]: Why is this being exported as BitcoinDaemon?

File: examples/basic_bitcoin/test/test.ts [hunk 3]
```diff
@@ -1,11 +1,27 @@
...
+const canisterId = getCanisterId(canisterName);
+
+let bitcoinDaemon: BitcoinDaemon;
+
+runTests(canisterName, () => {
+    beforeAll(async () => {
+        bitcoinDaemon = await startBitcoinDaemon();
+    });
+
+    afterAll(async () => {
```
Comment (lastmjs) [inline examples/basic_bitcoin/test/test.ts:19]: Does this need to be async?

File: examples/basic_bitcoin/test/tests.ts [hunk 4]
```diff
@@ -20,109 +21,102 @@ const FIRST_MINING_SESSION = 101;
...
+        });
+
+        it(`gets the balance of ${TO_ADDRESS} with /get-balance before ${TO_ADDRESS} receives any transactions`, async () => {
+            const balance = await getBalance(origin, TO_ADDRESS);
+
+            expect(balance).toBe(0n);
+        });
+
+        please(
+            `sends ${FIRST_AMOUNT_SENT} from canister to ${TO_ADDRESS} with /send`,
```
Comment (lastmjs) [inline examples/basic_bitcoin/test/tests.ts:118]: Why is this a please and not a test?

File: examples/bitcoin_psbt/package.json [hunk 5]
```diff
@@ -10,10 +10,10 @@
         "getrawtransaction": ".bitcoin/bin/bitcoin-cli -conf=$(pwd)/.bitcoin.conf getrawtransaction $npm_config_txid",
         "help": ".bitcoin/bin/bitcoin-cli -conf=$(pwd)/.bitcoin.conf help",
         "premanualtest": "ts-node --transpile-only --ignore=false test/pretest.ts",
-        "manualtest": "ts-node --transpile-only --ignore=false test/test.ts",
+        "manualtest": "jest test/manual_test.ts test/manual.test.ts",
```
Comment (lastmjs) [inline examples/bitcoin_psbt/package.json:13]: why are command-line arguments being passed into jest here?

File: examples/query/test/tests.ts [hunk 6]
```diff
@@ -1,20 +1,15 @@
...
 
-export function getTests(query_canister: ActorSubclass<_SERVICE>): Test[] {
-    return [
-        {
-            name: 'query',
-            test: async () => {
-                const result = await query_canister.simpleQuery();
+export function getTests(queryCanister: ActorSubclass<_SERVICE>): Test {
+    return () => {
+        it('makes a simple query calls', async () => {
```
Comment (lastmjs) [inline examples/query/test/tests.ts:9]: call instead of calls

File: test/jest.ts [hunk 7]
```diff
@@ -0,0 +1,152 @@
...
+    getPreEmitDiagnostics,
+    parseJsonConfigFileContent,
+    readConfigFile,
+    sys
+} from 'typescript';
+
+export type Test = () => void;
+
+export function runTests(
+    canisterName: string,
```
Comment (lastmjs) [inline test/jest.ts:17]: Is this really necessary?

File: test/jest.ts [hunk 8]
```diff
@@ -0,0 +1,152 @@
...
+    tests: Test,
+    cwd: string = process.cwd()
+) {
+    const { shouldRunTests, shouldRunTypeChecks, shouldRunBenchmark } =
```
Comment (lastmjs) [inline test/jest.ts:21]: benchmarks instead of benchmark right?

File: test/jest.ts [hunk 9]
```diff
@@ -0,0 +1,152 @@
...
+        processEnvVars();
+    if (shouldRunTests) {
```
Comment (lastmjs) [inline test/jest.ts:23]: spaces between these kinds of statements

File: test/jest.ts [hunk 10]
```diff
@@ -0,0 +1,152 @@
...
+
+export function wait(name: string, delay: number) {
+    test(
+        `wait ${name}`,
+        async () => {
+            console.info(`Waiting: ${delay} milliseconds ${name}`);
+            await new Promise((resolve) => {
+                setTimeout(resolve, delay);
+            });
+            expect(true);
```
Comment (lastmjs) [inline test/jest.ts:46]: Is this expect necessary?

File: test/jest.ts [hunk 11]
```diff
@@ -0,0 +1,152 @@
...
+        },
+        delay > 2_500 ? delay * 2 : 5_000 // Calling this function takes a few more milliseconds than what is specified in the delay. Since we never want the delay to time out we will add an ample amount of extra time to the timeout. Since doubling a short delay may not give us enough extra time we will have a reasonable floor.
```
Comment (lastmjs) [inline test/jest.ts:48]: Is it just a few milliseconds always? Can't we just add a constant few milliseconds instead then?

File: test/jest.ts [hunk 12]
```diff
@@ -0,0 +1,152 @@
...
+}
+
+function runTypeCheck(projectDir: string): string[] {
+    const configPath = findConfigFile(
+        projectDir,
+        sys.fileExists,
+        'tsconfig.json'
+    );
+
+    if (!configPath) {
```
Comment (lastmjs) [inline test/jest.ts:89]: I'm not sure we should force a tsconfig.json

File: test/jest.ts [hunk 13]
```diff
@@ -0,0 +1,152 @@
...
+    };
+
+    const program = createProgram(parsedCommandLine.fileNames, compilerOptions);
+    const emitResult = program.emit();
+
+    const allDiagnostics = getPreEmitDiagnostics(program).concat(
+        emitResult.diagnostics
+    );
+
+    const errorMessages = allDiagnostics.map((diagnostic) => {
```
Comment (lastmjs) [inline test/jest.ts:113]: We could make this more declarative, it technically violates some of the standards we have of nesting too much and not pulling out into a function

File: test/jest.ts [hunk 14]
```diff
@@ -0,0 +1,152 @@
...
+
+    return errorMessages;
+}
+
+function processEnvVars(): {
+    shouldRunTests: boolean;
+    shouldRunTypeChecks: boolean;
+    shouldRunBenchmark: boolean;
+} {
+    const shouldRunTests =
```
Comment (lastmjs) [inline test/jest.ts:137]: I think we should strictly check for the string true no these

File: test/jest.ts [hunk 15]
```diff
@@ -0,0 +1,152 @@
...
+        process.env.AZLE_INTEGRATION_TEST_RUN_TESTS !== undefined;
+    const shouldRunTypeChecks =
+        process.env.AZLE_INTEGRATION_TEST_RUN_TYPE_CHECKS !== undefined;
+    const shouldRunBenchmark =
+        process.env.AZLE_INTEGRATION_TEST_RUN_BENCHMARKS !== undefined;
+
+    return {
+        shouldRunTests:
+            shouldRunTests || (!shouldRunTypeChecks && !shouldRunBenchmark),
```
Comment (lastmjs) [inline test/jest.ts:146]: can you abstract this second condition and simply put it up into the variables above?

File: examples/management_canister/test/tests.ts [hunk 16]
```diff
@@ -1,321 +1,244 @@
...
+                canister_id: canisterId,
+                num_requested_changes: [50n]
+            });
+
+            expect(canisterInfo.total_num_changes).toBe(6n);
+            expect(canisterInfo.recent_changes).toHaveLength(6);
+            expect(canisterInfo.module_hash).toHaveLength(0);
+            expect(canisterInfo.controllers).toHaveLength(1);
+        });
+        it('executes delete_canister from the management canister', async () => {
```
Comment (lastmjs) [inline examples/management_canister/test/tests.ts:229]: Can we have linebreaks between its?

**Review by lastmjs — CHANGES_REQUESTED** [review]

File: examples/apollo_server/test/test.ts [hunk 17]
```diff
@@ -1,8 +1,10 @@
 import { getCanisterId } from 'azle/dfx';
-import { runTests } from 'azle/test';
+import { runTests } from 'azle/test/jest';
```
[the remark this answers is not in the document]
Comment (bdemann) [reply examples/apollo_server/test/test.ts]: yes we will do it later, its in the main issue

File: examples/async_await/jest.config.js [hunk 18]
```diff
@@ -0,0 +1,10 @@
+/** @type {import('ts-jest').JestConfigWithTsJest} */
+module.exports = {
+    preset: 'ts-jest',
+    testEnvironment: 'node',
+    testTimeout: 10_000,
```
[the remark this answers is not in the document]
Comment (bdemann) [reply examples/async_await/jest.config.js]: lets in all cases move it out of the config and onto the individual tests and round up to the nearest 10_000

File: examples/audio_and_video/test/test.ts [hunk 19]
```diff
@@ -1,8 +1,8 @@
 import { getCanisterId } from 'azle/dfx';
-import { runTests } from 'azle/test';
+import { runTests } from 'azle/test/jest';
 
 import { getTests } from './tests';
 
 const canisterId = getCanisterId('backend');
 
-runTests(getTests(canisterId));
+runTests('audio_and_video', getTests(canisterId));
```
[the remark this answers is not in the document]
Comment (bdemann) [reply examples/audio_and_video/test/test.ts]: it doesn't matter, we are getting rid of it

File: test/jest.ts (hunk 7 above)
Comment (bdemann) [reply test/jest.ts:17]: apparently not

File: examples/apollo_server/test/test.ts [hunk 20]
```diff
@@ -1,8 +1,10 @@
...
 
 import { getTests } from './tests';
 
-const canisterId = getCanisterId('apollo_server');
+const canisterName = 'apollo_server';
 
-runTests(getTests(canisterId));
+const canisterId = getCanisterId(canisterName);
+
+runTests(canisterName, getTests(canisterId));
```
[the remark this answers is not in the document]
Comment (bdemann) [reply examples/apollo_server/test/test.ts]: done

File: examples/autoreload/test/tests.ts
[the remark this answers is not in the document]
Comment (bdemann) [reply examples/autoreload/test/tests.ts:1]: yeah lets figure this out

File: test/jest.ts [hunk 21]
```diff
@@ -0,0 +1,152 @@
...
+        `it ${name}`,
+        async () => {
+            console.info(`Testing: ${name}`);
+            await fn();
+        },
+        timeout
+    );
+}
+
+function runTypeCheck(projectDir: string): string[] {
```
Comment (lastmjs) [inline test/jest.ts:82]: It would be nice to see the actual type errors, and have input as close to the typecsript commandline output as possible

**Review by lastmjs — CHANGES_REQUESTED** [review]

File: examples/autoreload/test/server_ts.ts
[the remark this answers is not in the document]
Comment (bdemann) [reply examples/autoreload/test/server_ts.ts:1]: done

File: examples/basic_bitcoin/test/bitcoin_daemon.ts (hunk 2 above)
Comment (bdemann) [reply examples/basic_bitcoin/test/bitcoin_daemon.ts:4]: to make it nicer to read even though it's slightly deceptive

File: examples/basic_bitcoin/test/test.ts (hunk 3 above)
Comment (bdemann) [reply examples/basic_bitcoin/test/test.ts:19]: no, done

File: examples/bitcoin_psbt/package.json (hunk 5 above)
Comment (bdemann) [reply examples/bitcoin_psbt/package.json:13]: to limit which tests are run

File: test/jest.ts (hunk 8 above)
Comment (bdemann) [reply test/jest.ts:21]: right

File: test/jest.ts (hunk 9 above)
Comment (bdemann) [reply test/jest.ts:23]: done

File: test/jest.ts (hunk 10 above)
Comment (bdemann) [reply test/jest.ts:46]: done

File: test/jest.ts (hunk 11 above)
Comment (bdemann) [reply test/jest.ts:48]: done

File: test/jest.ts (hunk 14 above)
Comment (bdemann) [reply test/jest.ts:137]: absolutely

File: test/jest.ts (hunk 15 above)
Comment (bdemann) [reply test/jest.ts:146]: done

File: examples/apollo_server/test/tests.ts [hunk 22]
```diff
@@ -1,74 +1,62 @@
 import * as dns from 'node:dns';
```
[the remark this answers is not in the document]
Comment (bdemann) [reply examples/apollo_server/test/tests.ts]: done

File: examples/async_await/package.json [hunk 23]
```diff
@@ -1,14 +1,16 @@
 {
     "scripts": {
         "pre_tests": "ts-node --transpile-only --ignore=false test/pretest.ts",
```
[the remark this answers is not in the document]
Comment (bdemann) [reply examples/async_await/package.json]: We decided to hold off on this for now. Here is an issue https://github.com/demergent-labs/azle/issues/1836

File: examples/async_await/test/tests.ts [hunk 24]
```diff
@@ -1,54 +1,33 @@
...
-    ];
+export function getTests(async_await_canister: ActorSubclass<_SERVICE>): Test {
+    return () => {
+        it('gets randomness when awaited once', async () => {
+            const result = await async_await_canister.getRandomnessDirectly();
+
+            expect(result).toHaveLength(32);
+        });
+
+        it('gets randomness even when there is multiple levels of awaits', async () => {
```
[the remark this answers is not in the document]
Comment (bdemann) [reply examples/async_await/test/tests.ts]: moved to the correct pr

File: examples/query/test/tests.ts (hunk 6 above)
Comment (bdemann) [reply examples/query/test/tests.ts:9]: done

File: test/jest.ts (hunk 7 above)
Comment (bdemann) [reply test/jest.ts:17]: done

File: test/jest.ts (hunk 8 above)
Comment (bdemann) [reply test/jest.ts:21]: done

File: examples/management_canister/test/tests.ts (hunk 16 above)
Comment (bdemann) [reply examples/management_canister/test/tests.ts:229]: absolutely, sorry I missed that

File: examples/async_await/jest.config.js (hunk 18 above)
[the remark this answers is not in the document]
Comment (bdemann) [reply examples/async_await/jest.config.js]: done

File: test/jest.ts (hunk 12 above)
Comment (bdemann) [reply test/jest.ts]: we are just using exec sync now

File: test/jest.ts (hunk 13 above)
Comment (bdemann) [reply test/jest.ts]: we are just using exec sync now

File: examples/management_canister/test/tests.ts (hunk 16 above)
Comment (bdemann) [reply examples/management_canister/test/tests.ts:232]: done

File: test/jest.ts (hunk 21 above)
Comment (bdemann) [reply test/jest.ts]: we are just using exec sync now

**Pull request merged**